File: core/src/main/java/com/sequenceiq/cloudbreak/service/eventbus/AbstractCloudPersisterService.java
package com.sequenceiq.cloudbreak.service.eventbus;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.annotation.PostConstruct;
import javax.inject.Inject;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.core.convert.ConversionService;
import org.springframework.data.repository.CrudRepository;
import org.springframework.util.CollectionUtils;

import com.sequenceiq.cloudbreak.cloud.service.Persister;
import com.sequenceiq.cloudbreak.repository.EntityType;

public abstract class AbstractCloudPersisterService<T> implements Persister<T> {

    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractCloudPersisterService.class);

    @Inject
    private List<CrudRepository> repositoryList;

    @Inject
    @Qualifier("conversionService")
    private ConversionService conversionService;

    private Map<Class, CrudRepository> repositoryMap = new HashMap<>();

    @Override
    public abstract T persist(T data);

    @Override
    public abstract T update(T data);

    @Override
    public abstract T retrieve(T data);

    @PostConstruct
    public void checkRepoMap() {
        if (CollectionUtils.isEmpty(repositoryList)) {
            throw new IllegalStateException("No repositories provided!");
        } else {
            fillRepositoryMap();
        }
    }

    protected ConversionService getConversionService() {
        return conversionService;
    }

    private void fillRepositoryMap() {
        for (CrudRepository repo : repositoryList) {
            repositoryMap.put(getEntityClassForRepository(repo), repo);
        }
    }

    private Class getEntityClassForRepository(CrudRepository repo) {
        Class<?> originalInterface = repo.getClass().getInterfaces()[0];
        EntityType annotation = originalInterface.getAnnotation(EntityType.class);
        if (annotation == null) {
            throw new IllegalStateException("Entity class is not specified for repository: " + originalInterface.getSimpleName());
        }
        return annotation.entityClass();
    }

    protected <T> T getRepositoryForEntity(Class clazz) {
        T repo = (T) repositoryMap.get(clazz);
        if (repo == null) {
            throw new IllegalStateException("No repository found for the entityClass:" + clazz);
        }
        return repo;
    }
}

File: core/src/main/java/com/sequenceiq/cloudbreak/service/eventbus/AbstractCloudPersisterService.kt
package com.sequenceiq.cloudbreak.service.eventbus

import java.util.HashMap

import javax.annotation.PostConstruct
import javax.inject.Inject

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.core.convert.ConversionService
import org.springframework.data.repository.CrudRepository
import org.springframework.util.CollectionUtils

import com.sequenceiq.cloudbreak.cloud.service.Persister
import com.sequenceiq.cloudbreak.repository.EntityType

abstract class AbstractCloudPersisterService<T> : Persister<T> {

    @Inject
    private val repositoryList: List<CrudRepository<Any, Serializable>>? = null

    @Inject
    @Qualifier("conversionService")
    protected val conversionService: ConversionService? = null

    private val repositoryMap = HashMap<Class<Any>, CrudRepository<Any, Serializable>>()

    abstract override fun persist(data: T): T

    abstract override fun update(data: T): T

    abstract override fun retrieve(data: T): T

    @PostConstruct
    fun checkRepoMap() {
        if (CollectionUtils.isEmpty(repositoryList)) {
            throw IllegalStateException("No repositories provided!")
        } else {
            fillRepositoryMap()
        }
    }

    private fun fillRepositoryMap() {
        for (repo in repositoryList!!) {
            repositoryMap.put(getEntityClassForRepository(repo), repo)
        }
    }

    private fun getEntityClassForRepository(repo: CrudRepository<Any, Serializable>): Class<Any> {
        val originalInterface = repo.javaClass.getInterfaces()[0]
        val annotation = originalInterface.getAnnotation(EntityType::class.java) ?: throw IllegalStateException("Entity class is not specified for repository: " + originalInterface.getSimpleName())
        return annotation!!.entityClass()
    }

    protected fun <T> getRepositoryForEntity(clazz: Class<Any>): T {
        val repo = repositoryMap[clazz] as T ?: throw IllegalStateException("No repository found for the entityClass:" + clazz)
        return repo
    }

    companion object {

        private val LOGGER = LoggerFactory.getLogger(AbstractCloudPersisterService<Any>::class.java)
    }
}
